dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,346 @@ module Dhall.LSP.Handlers where
+  txt <- readUri uri
+  mError <- lift $ gets (Map.lookup uri . view errors)
+  let isHovered (Diagnosis _ (Just (Range left right)) _) =
+        left <= (line,col) && (line,col) <= right
+      isHovered _ = False
+
+      hoverFromDiagnosis (Diagnosis _ (Just (Range left right)) diagnosis) =
+        let _range = Just $ J.Range (uncurry J.Position left)
+                                    (uncurry J.Position right)
+            encodedDiag = URI.encode (Text.unpack diagnosis)
+            command = "[Explain error](dhall-explain:?"
+                        <> Text.pack encodedDiag <> " )"
+            _contents = J.List [J.PlainString command]
+        in Just J.Hover { .. }
+      hoverFromDiagnosis _ = Nothing
+
+      mHover = do err <- mError
+                  explanation <- explain txt err
+                  True <- return $ isHovered explanation
+                  hoverFromDiagnosis explanation
+  lspRespond LSP.RspHover request mHover
+
+hoverType :: J.HoverRequest -> HandlerM ()
+hoverType request = do
+  let uri = request ^. J.params . J.textDocument . J.uri
+      (J.Position line col) = request ^. (J.params . J.position)
+  txt <- readUri uri
+  expr <- loadFile uri
+  (welltyped, _) <- case typecheck expr of
+    Left _ -> throwE (Info, "Can't infer type; code does not type-check.")
+    Right wt -> return wt
+  case typeAt (line,col) welltyped of
+    Left err -> throwE (Error, Text.pack err)
+    Right typ ->
+      let _range = fmap (rangeToJSON . sanitiseRange txt . rangeFromDhall)
+                        (srcAt (line,col) (fromWellTyped welltyped))
+          _contents = J.List [J.PlainString (pretty typ)]
+          hover = J.Hover{..}
+      in lspRespond LSP.RspHover request (Just hover)
+
+hoverHandler :: J.HoverRequest -> HandlerM ()
+hoverHandler request = do
+  let uri = request ^. J.params . J.textDocument . J.uri
+  errorMap <- lift $ gets (view errors)
+  case Map.lookup uri errorMap of
+    Nothing -> hoverType request
+    _ -> hoverExplain request
+
+
+diagnosticsHandler :: J.Uri -> HandlerM ()
+diagnosticsHandler uri = do
+  txt <- readUri uri
+  fileIdentifier <- fileIdentifierFromUri uri
+  -- make sure we don't keep a stale version around
+  lift $ modify (over importCache (invalidate fileIdentifier))
+  cache <- lift $ gets (view importCache)
+
+  errs <- flip catchE (return . Just) $ do
+      expr <- case parse txt of
+        Right e -> return e
+        Left err -> throwE err
+      loaded <- liftIO $ load fileIdentifier expr cache
+      (cache', expr') <- case loaded of
+        Right x -> return x
+        Left err -> throwE err
+      welltyped <- case typecheck expr' of
+        Right (wt, _typ) -> return wt
+        Left err -> throwE err
+      let normal = normalize welltyped
+      -- cache the new expression
+      lift $ modify (set importCache (cacheExpr fileIdentifier normal cache'))
+      return Nothing
+
+  let suggestions =
+        case parse txt of
```

Comment:
It looks like this function parses text twice.  It might be possible to parse the text once and then share the result between the two use sites.

Reply:
I believe parsing to be cheap compared to typechecking, so why complicate the code? :)

Reply:
I'll leave this for the time being. Might come back to it when the typechecker gets fast enough that this could make a difference.